test/Microsoft.AspNetCore.JsonPatch.Test/WriteOnceDynamicTestObject.cs

```diff
@@ -0,0 +1,117 @@
+    /// </para>
+    /// </remarks>
+    public class WriteOnceDynamicTestObject : DynamicObject
+    {
+        private Dictionary<string, object> _dictionary = new Dictionary<string, object>();
+
+        public object this[string key] { get => ((IDictionary<string, object>)_dictionary)[key]; set => SetValueForKey(key, value); }
+
+        public ICollection<string> Keys => ((IDictionary<string, object>)_dictionary).Keys;
+
+        public ICollection<object> Values => ((IDictionary<string, object>)_dictionary).Values;
+
+        public int Count => ((IDictionary<string, object>)_dictionary).Count;
+
+        public bool IsReadOnly => ((IDictionary<string, object>)_dictionary).IsReadOnly;
+
+        public void Add(string key, object value)
+        {
+            SetValueForKey(key, value);
+        }
+
+        public void Add(KeyValuePair<string, object> item)
+        {
+            SetValueForKey(item.Key, item.Value);
+        }
+
+        public void Clear()
+        {
+            ((IDictionary<string, object>)_dictionary).Clear();
+        }
+
+        public bool Contains(KeyValuePair<string, object> item)
+        {
+            return ((IDictionary<string, object>)_dictionary).Contains(item);
+        }
+
+        public bool ContainsKey(string key)
+        {
+            return ((IDictionary<string, object>)_dictionary).ContainsKey(key);
+        }
+
+        public void CopyTo(KeyValuePair<string, object>[] array, int arrayIndex)
+        {
+            ((IDictionary<string, object>)_dictionary).CopyTo(array, arrayIndex);
+        }
+
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+        {
+            return ((IDictionary<string, object>)_dictionary).GetEnumerator();
+        }
+
+        public bool Remove(string key)
+        {
+            return ((IDictionary<string, object>)_dictionary).Remove(key);
+        }
+
+        public bool Remove(KeyValuePair<string, object> item)
+        {
+            return ((IDictionary<string, object>)_dictionary).Remove(item);
+        }
+
+        public bool TryGetValue(string key, out object value)
+        {
+            return ((IDictionary<string, object>)_dictionary).TryGetValue(key, out value);
+        }
+
+        public override bool TryGetMember(GetMemberBinder binder, out object result)
+        {
+            var name = binder.Name;
+
+            return TryGetValue(name, out result);
+        }
+
+        public override bool TrySetMember(SetMemberBinder binder, object value)
+        {
+            SetValueForKey(binder.Name, value);
+
+            return true;
+        }
+
+        private void SetValueForKey(string key, object value)
+        {
+            if (value == null)
+            {
+                _dictionary.Remove(key);
+                return;
+            }
+
+            if (_dictionary.ContainsKey(key))
+            {
+                throw new ArgumentException($"Value for {key} already exists");
+            }
```

Comment:
Why is this an exception? Curious why the new value is not set? `_dictionary[binder.Name] = value`

https://msdn.microsoft.com/en-us/library/zyxt2e2h(v=vs.110).aspx

Reply:
The purpose of this class (and the reason it's called "WriteOnce") is to assert that dictionary values are not overwritten, at least not without being removed first. From the documentation I added at the top:

> This is done by asserting that  no value exists for a particular key before setting its value. To replace the value for a key, the key must first be removed, and then re-added with the new value.

I modelled it similarly to the logic I use in my own project, which is how I discovered the issue.

Reply:
This seems like a very specific implementation requirement. The current `Replace` operation supports the generic scenario.

Reply:
> This seems like a very specific implementation requirement. 

Admittedly yeah, there's few cases where the distinction would be noticeable.

Still:
 - it's what the spec dictates. There should be no difference between doing a replace operation, and doing a remove operation followed by an add operation. I've shown one case where the current implementation doesn't adhere to that. Possibly others have encountered more.
 - It's the way you had it implemented originally. Changing it now constitutes a breaking change. Technically that's fine since you're a whole major version ahead at this point, but it seems like a gratuitous one.
 - Importantly, this change should not break any existing functionality.